// @nestjs/azure-database/dist/cosmos-db/cosmos-db.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var AzureCosmosDbModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureCosmosDbModule = void 0;
const common_1 = require("@nestjs/common");
const cosmos_db_core_module_1 = require("./cosmos-db-core.module");
const cosmos_db_providers_1 = require("./cosmos-db.providers");
let AzureCosmosDbModule = AzureCosmosDbModule_1 = class AzureCosmosDbModule {
    static forRoot(options) {
        return {
            module: AzureCosmosDbModule_1,
            imports: [cosmos_db_core_module_1.AzureCosmosDbCoreModule.forRoot(options)],
        };
    }
    static forRootAsync(options) {
        return {
            module: AzureCosmosDbModule_1,
            imports: [cosmos_db_core_module_1.AzureCosmosDbCoreModule.forRootAsync(options)],
        };
    }
    static forFeature(models = [], connectionName) {
        const providers = (0, cosmos_db_providers_1.createAzureCosmosDbProviders)(connectionName, models);
        return {
            module: AzureCosmosDbModule_1,
            providers,
            exports: providers,
        };
    }
};
AzureCosmosDbModule = AzureCosmosDbModule_1 = __decorate([
    (0, common_1.Module)({})
], AzureCosmosDbModule);
exports.AzureCosmosDbModule = AzureCosmosDbModule;
