// @nestjs/azure-database/dist/cosmos-db/cosmos-db-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var AzureCosmosDbCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureCosmosDbCoreModule = void 0;
const cosmos_1 = require("@azure/cosmos");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const rxjs_1 = require("rxjs");
const cosmos_db_constants_1 = require("./cosmos-db.constants");
const cosmos_db_utils_1 = require("./cosmos-db.utils");
let AzureCosmosDbCoreModule = AzureCosmosDbCoreModule_1 = class AzureCosmosDbCoreModule {
    constructor(connectionName, moduleRef) {
        this.connectionName = connectionName;
        this.moduleRef = moduleRef;
    }
    static forRoot(options) {
        const { dbName, retryAttempts, retryDelay, connectionName } = options, cosmosDbOptions = __rest(options, ["dbName", "retryAttempts", "retryDelay", "connectionName"]);
        const cosmosConnectionName = (0, cosmos_db_utils_1.getConnectionToken)(connectionName);
        const cosmosConnectionNameProvider = {
            provide: cosmos_db_constants_1.COSMOS_DB_CONNECTION_NAME,
            useValue: cosmosConnectionName,
        };
        const connectionProvider = {
            provide: cosmosConnectionName,
            useFactory: () => __awaiter(this, void 0, void 0, function* () {
                return yield (0, rxjs_1.defer)(() => __awaiter(this, void 0, void 0, function* () {
                    const client = new cosmos_1.CosmosClient(cosmosDbOptions);
                    const dbResponse = yield client.databases.createIfNotExists({
                        id: dbName,
                    });
                    return dbResponse.database;
                }))
                    .pipe((0, cosmos_db_utils_1.handleRetry)(retryAttempts, retryDelay))
                    .toPromise();
            }),
        };
        return {
            module: AzureCosmosDbCoreModule_1,
            providers: [connectionProvider, cosmosConnectionNameProvider],
            exports: [connectionProvider],
        };
    }
    static forRootAsync(options) {
        const cosmosConnectionName = (0, cosmos_db_utils_1.getConnectionToken)(options.connectionName);
        const cosmosConnectionNameProvider = {
            provide: cosmos_db_constants_1.COSMOS_DB_CONNECTION_NAME,
            useValue: cosmosConnectionName,
        };
        const connectionProvider = {
            provide: cosmosConnectionName,
            useFactory: (cosmosModuleOptions) => __awaiter(this, void 0, void 0, function* () {
                const { dbName, retryAttempts, retryDelay, connectionName } = cosmosModuleOptions, cosmosOptions = __rest(cosmosModuleOptions, ["dbName", "retryAttempts", "retryDelay", "connectionName"]);
                return yield (0, rxjs_1.defer)(() => __awaiter(this, void 0, void 0, function* () {
                    const client = new cosmos_1.CosmosClient(cosmosOptions);
                    const dbResponse = yield client.databases.createIfNotExists({
                        id: dbName,
                    });
                    return dbResponse.database;
                }))
                    .pipe((0, cosmos_db_utils_1.handleRetry)(retryAttempts, retryDelay))
                    .toPromise();
            }),
            inject: [cosmos_db_constants_1.COSMOS_DB_MODULE_OPTIONS],
        };
        const asyncProviders = this.createAsyncProviders(options);
        return {
            module: AzureCosmosDbCoreModule_1,
            imports: options.imports,
            providers: [...asyncProviders, connectionProvider, cosmosConnectionNameProvider],
            exports: [connectionProvider],
        };
    }
    static createAsyncProviders(options) {
        if (options.useExisting || options.useFactory) {
            return [this.createAsyncOptionsProvider(options)];
        }
        const useClass = options.useClass;
        return [
            this.createAsyncOptionsProvider(options),
            {
                provide: useClass,
                useClass,
            },
        ];
    }
    static createAsyncOptionsProvider(options) {
        if (options.useFactory) {
            return {
                provide: cosmos_db_constants_1.COSMOS_DB_MODULE_OPTIONS,
                useFactory: options.useFactory,
                inject: options.inject || [],
            };
        }
        const inject = [(options.useClass || options.useExisting)];
        return {
            provide: cosmos_db_constants_1.COSMOS_DB_MODULE_OPTIONS,
            useFactory: (optionsFactory) => __awaiter(this, void 0, void 0, function* () { return yield optionsFactory.createAzureCosmosDbOptions(); }),
            inject,
        };
    }
};
AzureCosmosDbCoreModule = AzureCosmosDbCoreModule_1 = __decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({}),
    __param(0, (0, common_1.Inject)(cosmos_db_constants_1.COSMOS_DB_CONNECTION_NAME)),
    __metadata("design:paramtypes", [String, core_1.ModuleRef])
], AzureCosmosDbCoreModule);
exports.AzureCosmosDbCoreModule = AzureCosmosDbCoreModule;
